src/relayserver/TransactionManager.ts

```diff
@@ -0,0 +1,208 @@
+  receipt: TransactionReceipt
+  signedTx: PrefixedHexString
+}
+
+export interface SendTransactionDetails {
+  signer: Address
+  method?: any
+  destination: Address
+  value?: IntString
+  gasLimit?: IntString
+  gasPrice?: IntString
+}
+
+const pendingTransactionTimeout = 5 * 60 * 1000 // 5 minutes in milliseconds
+const confirmationsNeeded = 12
+const retryGasPriceFactor = 1.2
+const maxGasPrice = 100e9
+
+export class TransactionManager {
+  nonceMutex = new Mutex()
+  managerKeyManager: KeyManager
+  workersKeyManager: KeyManager
+  contractInteractor: ContractInteractor
+  nonces: Record<Address, number> = {}
+  txStoreManager: TxStoreManager
+
+  rawTxOptions!: TransactionOptions
+
+  constructor (contractInteractor: ContractInteractor, dependencies: ServerDependencies) {
+    this.contractInteractor = contractInteractor
+
+    this.txStoreManager = dependencies.txStoreManager
+    this.workersKeyManager = dependencies.workersKeyManager
+    this.managerKeyManager = dependencies.managerKeyManager
+
+    // todo: initialize nonces for all signers (currently one manager, one worker)
+    this.nonces[this.managerKeyManager.getAddress(0)] = 0
+    this.nonces[this.workersKeyManager.getAddress(0)] = 0
+  }
+
+  async _init (): Promise<void> {
+    await this.contractInteractor._init()
+    this.rawTxOptions = this.contractInteractor.getRawTxOptions()
+  }
+
+  async sendTransaction ({ signer, method, destination, value = '0x', gasLimit, gasPrice }: SendTransactionDetails): Promise<SignedTransactionDetails> {
+    const encodedCall = method?.encodeABI() ?? '0x'
+    const _gasPrice = parseInt(gasPrice ?? await this.contractInteractor.getGasPrice())
+    log.debug('gasPrice', _gasPrice)
+    log.debug('encodedCall', encodedCall)
+    const gas = parseInt(gasLimit ?? await method?.estimateGas({ from: signer }))
+    log.debug('gasLimit', gas)
+    log.debug('nonceMutex locked?', this.nonceMutex.isLocked())
+    const releaseMutex = await this.nonceMutex.acquire()
+    let signedTx
+    let storedTx: StoredTx
+    try {
+      const nonce = await this.pollNonce(signer)
+      log.debug('nonce', nonce)
+      const txToSign = new Transaction({
+        to: destination,
+        value: value,
+        gasLimit: gas,
+        gasPrice: _gasPrice,
+        data: Buffer.from(encodedCall.slice(2), 'hex'),
+        nonce
+      }, this.rawTxOptions)
+      log.trace('txToSign', txToSign)
+      // TODO omg! do not do this!
+      const keyManager = this.managerKeyManager.isSigner(signer) ? this.managerKeyManager : this.workersKeyManager
+      signedTx = keyManager.signTransaction(signer, txToSign)
+      storedTx = transactionToStoredTx(txToSign, signer, 1)
+      this.nonces[signer]++
+      await this.txStoreManager.putTx(storedTx, false)
+    } finally {
+      releaseMutex()
+    }
+    const receipt = await this.contractInteractor.sendSignedTransaction(signedTx)
+    log.debug('\ntxhash is', receipt.transactionHash)
```

Comment:
This should be info imo, logging tx hash for each transaction sent by server is reasonable.

Reply:
Will fix when working on the multiple workers

Reply:
fixed